enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -6252,83 +6335,650 @@ class AdjointGenerator
         if (!forwardMode)
           op = lookup(op, Builder2);
 
-        if (forwardMode) {
-          Value *args[] = {
-              /*sendbuf*/ shadow_sendbuf,
-              /*recvbuf*/ shadow_recvbuf,
-              /*count*/ count,
-              /*datatype*/ datatype,
-              /*op*/ op,
-              /*comm*/ comm,
-          };
+        auto getLLVMTypeFromMPIDatatype = [](Value *datatype) {
+          Type *Ty = nullptr;
+          if (Constant *C = dyn_cast<Constant>(datatype)) {
+            while (ConstantExpr *CE = dyn_cast<ConstantExpr>(C)) {
+              C = CE->getOperand(0);
+            }
+            // OpenMPI
+            if (auto GV = dyn_cast<GlobalVariable>(C)) {
+              auto name = GV->getName();
+              auto &C = GV->getContext();
+              if (name == "ompi_mpi_double") {
+                Ty = Type::getDoubleTy(C);
+              } else if (name == "ompi_mpi_float") {
+                Ty = Type::getFloatTy(C);
+              }
+            }
+            // MPICH
+            if (ConstantInt *CI = dyn_cast<ConstantInt>(C)) {
+              auto number = CI->getLimitedValue();
+              if (number == 1275070475) { // MPI_DOUBLE
+                Ty = Type::getDoubleTy(CI->getContext());
+              } else if (number == 1275069450) { // MPI_FLOAT
+                Ty = Type::getFloatTy(CI->getContext());
+              }
+            }
+          }
+          assert((Ty != nullptr) && "Unhandled MPI dataype");
+          return Ty;
+        };
 
-#if LLVM_VERSION_MAJOR >= 11
-          auto callval = call.getCalledOperand();
+        // If the primal op is min/max then replace it with the minloc/maxloc
+        // op in the primal code
+        CallInst *newNewCall;
+        Value *recvlocbuf;
+        if (opKind == MPI_Op_Kind::MIN || opKind == MPI_Op_Kind::MAX) {
```

Comment:
This is the block where I construct the replacement code for MPI_Allreduce when the op is min/max, and then replace the primal call.  The replacement code is constructed in a function (I took a similar approach as getOrInsertCheckedFree. Is this the best approach? Or is it better to generate the code in-place?